Sources/NIOCore/IPAddresses.swift

```diff
@@ -0,0 +1,447 @@
+
+    /// An IPv6 `IPAddress`.
+    case v6(IPv6Address)
+
+    /// A human-readable description of this `IPAddress`. Mostly useful for logging.
+    public var description: String {
+        switch self {
+        case .v4(_):
+            return "[IPv4]\(self.ipAddress)"
+        case .v6(_):
+            return "[IPv6]\(self.ipAddress)"
+        }
+    }
+    
+    /// Get the `IPAddress` as a string.
+    public var ipAddress: String {
+        switch self {
+        case .v4(let addr):
+            return addr.ipAddress
+        case .v6(let addr):
+            return addr.ipAddress
+        }
+    }
+
+    /// Creates a `IPAddress` directly out of UInt8 Tuple for IPv4.
+    public init(_ ipv4BytesTuple: IPv4BytesTuple) {
+        self = .v4(IPv4Address(address: .init(ipv4BytesTuple)))
+    }
+    
+    /// Creates a `IPAddress` directly out of UInt8 Tuple for IPv6.
+    public init(_ ipv6BytesTuple: IPv6BytesTuple) {
+        self = .v6(IPv6Address(address: .init(ipv6BytesTuple)))
+    }
+    
+    /// Creates a new `IPAddress` for the given string.
+    /// "d.d.d.d" with decimal values for IPv4 and "h:h:h:h:h:h:h:h" with hexadecimal values for IPv6. Also allowing for shortened IPv6 representation with "::". Hybrid versions for IPv6 are not (yet) supported.
+    ///
+    /// - parameters:
+    ///     - string: String representation of IPv4 or IPv6 Address
+    /// - returns: The `IPAddress` for the given string
+    /// - throws: May throw `IPAddressError.failedToParseIPString` if the string cannot be parsed to IPv4 or IPv6.
+    public init(string: String) throws {
+        do {
+            try self = .v4(.init(string: string))
+        } catch {
+            try self = .v6(.init(string: string))
+        }
+    }
+    
+    /// Creates a new `IPAddress` for the given bytes.
+    ///
+    /// - parameters:
+    ///     - bytes: Either 4 or 16 bytes representing the IPAddress value.
+    /// - returns: The `IPAddress` for the given string or `nil` if the string representation is not supported.
+    public init<Bytes: Collection>(packedBytes bytes: Bytes) throws where Bytes.Element == UInt8, Bytes.Index == Int {
+        switch bytes.count {
+        case 4: self = .v4(.init(packedBytes: bytes))
+        case 16: self = .v6(.init(packedBytes: bytes))
+        default:
+            throw IPAddressError.bytesArrayHasWrongLength(bytes.count)
+        }
+    }
+    
+    /// Creates a new `IPAddress` for the given libc IPv4 address.
+    ///
+    /// - parameters:
+    ///     - posixIPv4Address: libc ipv4 address.
+    public init(posixIPv4Address: in_addr) {
+        let uint8Bitmask: UInt32 = 0x000000FF
+        
+        let uint8AddressBytes: IPv4Bytes = .init((
+            UInt8((posixIPv4Address.s_addr >> 24) & uint8Bitmask),
+            UInt8((posixIPv4Address.s_addr >> 16) & uint8Bitmask),
+            UInt8((posixIPv4Address.s_addr >> 8) & uint8Bitmask),
+            UInt8(posixIPv4Address.s_addr & uint8Bitmask)
```

Comment:
Nit: we don't need the bit mask here, we can use `UInt8(truncatingIfNeeded:)` instead.